crates/host/src/wasmbus/providers/http_server/path.rs

```diff
@@ -0,0 +1,374 @@
+use core::net::SocketAddr;
+
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use anyhow::{bail, Context as _};
+use http::header::HOST;
+use http::uri::Scheme;
+use http::Uri;
+use http_body_util::BodyExt as _;
+use tokio::sync::RwLock;
+use tokio::task::JoinHandle;
+use tokio::time::Instant;
+use tracing::{debug, error, info_span, instrument, trace_span, warn, Instrument as _, Span};
+use wasmcloud_provider_sdk::{LinkConfig, LinkDeleteInfo};
+use wasmcloud_tracing::KeyValue;
+use wrpc_interface_http::ServeIncomingHandlerWasmtime as _;
+
+use crate::wasmbus::{Component, InvocationContext};
+
+use super::listen;
+
+/// This struct holds both the forward and reverse mappings for path-based routing
+/// so that they can be modified by just acquiring a single lock in the [`HttpServerProvider`]
+#[derive(Default)]
+pub(crate) struct Router {
+    /// Lookup from a path to the component ID that is handling that path
+    pub(crate) paths: HashMap<Arc<str>, Arc<str>>,
+    /// Reverse lookup to find the path for a (component,link_name) pair
+    pub(crate) components: HashMap<(Arc<str>, Arc<str>), Arc<str>>,
+}
+
+pub(crate) struct Provider {
+    /// Handle to the server task
+    pub(crate) handle: JoinHandle<()>,
+    /// Struct that holds the routing information based on path/component_id
+    pub(crate) path_router: Arc<RwLock<Router>>,
+}
+
+// Implementations of put and delete link are done in the `impl Provider` block to aid in testing
+impl wasmcloud_provider_sdk::Provider for Provider {
+    #[instrument(level = "debug", skip_all)]
+    async fn receive_link_config_as_source(&self, link: LinkConfig<'_>) -> anyhow::Result<()> {
+        self.put_link(link.target_id, link.link_name, link.config)
+            .await
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn delete_link_as_source(&self, info: impl LinkDeleteInfo) -> anyhow::Result<()> {
+        self.delete_link(
+            info.get_source_id(),
+            info.get_target_id(),
+            info.get_link_name(),
+        )
+        .await
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    async fn shutdown(&self) -> anyhow::Result<()> {
+        self.handle.abort();
```

Comment:
doesn't this also leak the task? Doesn't look like `delete_link_as_source` is guaranteed to be called if this is called